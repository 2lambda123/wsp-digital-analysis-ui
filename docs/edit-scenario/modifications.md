# Modification details

At the top of the modification details panel, there are multiple actions for each modification, including:
<br><span class="ui-icon"><i class="fa fa-pencil"></i>Rename</span>
<br><span class="ui-icon"><i class="fa fa-copy"></i>Copy</span>
<br><span class="ui-icon"><i class="fa fa-trash"></i>Delete</span>

Clicking the small blue back arrow (<span class="ui-icon"><i class="fa fa-chevron-left"></i>Modifications</span>) saves your changes and take you back to the list of all modifications.  Changes are also automatically saved every 10 seconds.  Each type of modification has additional options available, as detailed below.

## Add trips

The add trips modification allows you to add new [trip patterns](../glossary.html#trip-pattern) to your transport scenario. Generally
one add-trip modification is used to model a new route, adding all of the trip patterns on that route. After creating the modification you will see this view:

<figure>
<img src="../img/new-add-trip-pattern.png" alt="a new add trips modification" />
</figure>

To create an alignment for the modification (or to edit the alignment you've previously created), click
<br><span class="btn btn-warning"><i class="fa fa-pencil"></i> Edit route geometry</span>

This will activate editing mode.  In editing mode, click once on the map to place the first stop, then again to place the second stop, and so on. If you click on an existing stop (indicated by a small gray circle), the icon for the new stop will be black and the new transit service will stop at that existing stop. If you click in a place where there is not an existing stop, a new stop (in blue) will be created.  

To insert a stop into the middle of an alignment, click any part of the alignment.  Once created, any stop can be dragged to move it.  Clicking on a stop gives you the option to delete it, or convert it to a _control point_ through which the route will pass without stopping. Similarly, control points can be converted back to stops or deleted by clicking on them.  It's important to get the alignment approximately correct so that the length of each segment is correct when used to calculate the travel times between stops.
<figure>
<img src="../img/make-stop.png" alt="make a new stop" />
</figure>

Available options while editing an alignment include:
* **Auto-create stops**: Whether stops should be created automatically at a specified interval along the route.
* **Bidirectional**: If this checkbox is checked, vehicles will travel in both directions along the described geometry. If it is left unchecked, vehicles will only travel in the direction the line is drawn (which can be useful when there are couplets or other aspects of the route that don't follow the same alignment in both directions). You can choose whether stops should be created automatically at a specified interval along the route, or if you will create all stops manually.
* **Follow streets**: Make the route follow the streets. This only applies to segments that are actively being edited, and will not cause already drawn segments to follow the streets, allowing you to draw part of a route on street and part off-street.

Once you have created an alignment, you need to specify when the routes run using a [simplified timetable](timetable.html). You can do this by [copying a previously created timetable](timetable.html#copying-timetables), or by clicking:
<br><span class="btn btn-success"><i class="fa fa-plus"></i> Add timetable</span>


## Adjust dwell time
You may also want to adjust the dwell time along a route or at a particular stop, for example to model the effects of off-board fare collection, or the effects of increasing ridership at a particular
stop. As with the remove-stops modification, you can select a feed, route and optionally patterns. You can then use the map to select the affected stops (if you skip this step, all stops will have their
dwell times adjusted). You can then choose to either enter a new dwell time (in seconds), or scale the existing dwell times (for instance, entering 2 would double exiting dwell times).

Unfortunately, many source GTFS feeds do not contain any dwell time, meaning that this modification is of limited usefulness in some cases.

<figure>
<img src="../img/adjust-dwell-time.png" alt="Adjusting the dwell time at particular stops on a route" />
</figure>

## Adjust frequency

Often a scenario will include frequency changes to existing lines. We support this using the adjust frequency/convert to frequency modification. First, create a "convert to frequency" modification, and choose the feed and route you want to adjust:

<figure>
<img src="../img/new-change-frequency.png" alt="Selecting the route on which to change frequencies" />
</figure>

You then create any number of frequency entries using [simple timetables](timetable.html). Typically, there will be at least two entries (one for each direction). Each frequency entry will follow one template trip, which determines the stop-to-stop travel times of the trips that are created with the given frequency.

You then [add details to each frequency entry](timetable.html). Repeat this process until you have created frequency entries for all the service you want to retain, in both directions.

You can choose to remove all existing trips on the route (the default) or choose to retain trips outside the time windows in which you specify frequencies, which is useful when you are changing the frequency for only part of the day (e.g. increased weekend frequency) and want to retain the existing scheduled service at other times. This is controlled using the "Retain existing scheduled trips at times without new frequencies specified" checkbox.



## Adjust speed

Sometimes you will want to adjust the speed on a route, or a portion of it (for instance due to the addition of dedicated bus lanes, or an application of transit signal priority).
As before, you will create the modification and select a feed, routes and patterns. If you want to modify only part of the route, you can use similar selection tools to those used in the
remove stops (with new selection, add and remove buttons). The difference is that you are now selecting [hops](../glossary.html#hops). The selected segment will be shown on the map in purple.
Finally, enter a scale, which is the factor to multiply the speed by. For instance, if you enter 2, vehicles will travel twice as fast between stops (this modification does not affect dwell times;
to model changes in dwell time, see the adjust dwell time modification below).

This modification does not take into account the possibility of increased frequency due to more efficient
routes. However, it can be paired with a change frequency modification to model that scenario.

<figure>
<img src="../img/adjust-speed.png" alt="Adjusting the speed of a portion of a transit line" />
</figure>

## Remove stops

It is also possible to remove some of the stops from a route, while leaving the rest of the route untouched. To do this, create a new Remove Stops modification, and select a feed, route, and
patterns as you did when removing trips. You can then use the map to select which stops to remove. Under "Selection," click "new", "add to," or "remove from" to select new stops to remove, add to
the existing selection, or remove from the existing selection. Stops that are selected for removal are listed at the bottom of the modification, as well as being highlighted in red on the map.

You can also specify that you wish to remove a certain amount of time at each removed stop (in addition to the dwell time specified in the original feed, which is automatically removed). This could be used to account
for acceleration and deceleration, and can also be used when the original feed does not specify dwell time.
This feature is useful when testing the effects of stop consolidation.

When removing the beginning of a route, the dwell times at each stop are removed, as is any time
specified to be removed at each removed stop. Any remaining travel time is preserved as an offset
from the start of the trip in the original GTFS to the start of trips at the new first stop. This is
effectively as if the vehicles were leaving the original terminal at the same time but deadheading
past all of the removed stops.

This modification does not take into account the possibility of increased frequency due to more efficient
routes. However, it can be paired with a change frequency modification to model that scenario.

<figure>
<img src="../img/remove-stops.png"  alt="Remove stops" />
</figure>

## Remove trips

Another common modification is to remove trips. The most common use is to remove entire routes, but it is also possible to remove only specific patterns on a particular routes.
In order to remove trips, create a new Remove Trips modification, and select a GTFS feed, route, and optionally the trip pattern of the trips to be removed. All trips on this
route/pattern combination will be removed, and the route will be shown in red on the map. Note that the "active in variants" selector always specifies whether the modification is active,
so in this case implies that the trips will be removed from the selected variants.

<figure>
<img src="../img/remove-trips.png" alt="Remove trips" />
</figure>

## Reroute

It is also possible to add stops to a route (i.e. reroute the route). You can do this at the beginning or end of a route, or have a reroute in the middle.
First, create a Reroute modification, and select a route and patterns (as before). You then select the start and end stop of the reroute (i.e. where the route begins
and ends its deviation from the its existing pattern). Note that you will need to select patterns appropriately and create one reroute for each direction of the route;
a reroute can apply to multiple patterns in a single direction as long as they both contain the start and end stop in order. You select a stop by clicking the select stop
button, and then choosing the stop on the map. If you leave the from stop blank, the entire route up to the from stop will be rerouted. If you leave the to stop blank, the entire route
after that point will be rerouted. This can be used to extend routes, or to divert the ends of routes (e.g. to new rail stations).

<figure>
<img src="../img/reroute.png" alt="Rerouting" />
</figure>

You can then edit the alignment and the intermediate stops by clicking
<br><span class="btn btn-warning"><i class="fa fa-pencil"></i> Edit route geometry</span>

This activates editing mode, which is explained above for [adding trip patterns](#add-trips).

This modification is displayed on the map with the route being modified in gray, any segments that are being replaced in red, and the new segments in blue.
