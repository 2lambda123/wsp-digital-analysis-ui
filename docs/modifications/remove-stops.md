It is also possible to remove some of the stops from a route, while leaving the rest of the route untouched. To do this, create a new Remove Stops modification, and select a feed, route, and
patterns as you did when removing trips. You can then use the map to select which stops to remove. Under "Selection," click "new", "add to," or "remove from" to select new stops to remove, add to
the existing selection, or remove from the existing selection. Stops that are selected for removal are listed at the bottom of the modification, as well as being highlighted in red on the map.

You can also specify that you wish to remove a certain amount of time at each removed stop (in addition to the
dwell time specified in the original feed, which is automatically removed). This could be used to account
for acceleration and deceleration, and can also be used when the original feed does not specify dwell time.
This feature is useful when testing the effects of stop consolidation.

When removing the beginning of a route, the dwell times at each stop are removed, as is any time
specified to be removed at each removed stop. Any remaining travel time is preserved as an offset
from the start of the trip in the original GTFS to the start of trips at the new first stop. This is
effectively as if the vehicles were leaving the original terminal at the same time but deadheading
past all of the removed stops.

This modification does not take into account the possibility of increased frequency due to more efficient
routes. However, it can be paired withe a change frequency modification to model that scenario.

<img src="../../img/remove-stops.png"  alt="Remove stops" />
