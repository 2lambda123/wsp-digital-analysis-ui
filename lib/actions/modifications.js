// @flow
import fetch from '@conveyal/woonerf/fetch'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'

import {lockUiWithError} from './'
import copyModification from '../utils/copy-modification'
import copyModifications from '../utils/copy-modifications'
import getFeedsRoutesAndStops from './get-feeds-routes-and-stops'
import {
  create as createModification,
  formatForServer,
  isValid
} from '../utils/modification'

import type {Modification} from '../types'

type GetState = () => any

/**
 * TODO: Move to a server side API call?
 * POST the modification to `/api/modification/copy` ?
 */
export function copy ({modification}: {modification: Modification}) {
  return fetch({
    url: '/api/modification',
    options: {
      body: copyModification(modification),
      method: 'post'
    },
    next (error, response) {
      if (!error) {
        const newModification = response.value
        return [
          setLocally(newModification),
          push(
            `/scenarios/${modification.scenarioId}/modifications/${newModification._id}`
          )
        ]
      }
    }
  })
}

/**
 * TODO: Move to a server side API call?
 * POST a scenario id to `/api/scenario/copy-modifications`?
 */
export function copyFromScenario ({fromScenarioId, toScenarioId, variants}: {
  fromScenarioId: string,
  toScenarioId: string,
  variants: boolean[]
}) {
  return fetch({
    url: `/api/scenario/${fromScenarioId}/modifications`,
    next: (error, response) =>
      !error &&
      copyModifications(
        response.value,
        toScenarioId,
        variants
      ).map(newModification => set(newModification))
  })
}

export const create = ({name, type}: {
  name: string,
  type: string
}) => (dispatch: Dispatch, getState: GetState) => {
  const {scenario} = getState()
  const {currentScenario, feeds} = scenario
  const scenarioId = currentScenario._id

  return dispatch(
    fetch({
      url: '/api/modification',
      options: {
        body: createModification({
          feedId: feeds.length > 0 ? feeds[0].id : '',
          name,
          scenarioId,
          type,
          variants: currentScenario.variants.map(v => true)
        }),
        method: 'post'
      },
      next (error, response) {
        if (!error) {
          const modification = response.value
          return [
            setLocally(modification),
            push(`/scenarios/${scenarioId}/modifications/${modification._id}`)
          ]
        }
      }
    })
  )
}

export const update = (modification: Modification) =>
  (dispatch: Dispatch, getState: GetState) => {
    const {scenario} = getState()
    const {currentScenario} = scenario

    return dispatch(
      isValid(modification)
        ? [
          set(modification),
          getFeedsRoutesAndStops({
            bundleId: currentScenario.bundleId,
            modifications: [modification]
          })
        ]
        : [lockUiWithError('Modification is not valid after operation')]
    )
  }

const deleteLocally = createAction('delete modification')
const deleteOnServer = _id =>
  fetch({
    url: `/api/modification/${_id}`,
    options: {
      method: 'delete'
    },
    next () {}
  })
export const deleteModification = (_id: string) =>
  (dispatch: Dispatch, getState: GetState) => {
    const {scenario} = getState()
    const modifications = scenario ? scenario.modifications : []
    return dispatch([
      deleteLocally(_id),
      deleteOnServer(_id),
      clearReferencesToModification({_id, modifications})
    ])
  }

export function getForScenario ({bundleId, scenarioId}: {
  bundleId: string,
  scenarioId: string
}) {
  return fetch({
    url: `/api/scenario/${scenarioId}/modifications`,
    next (error, response) {
      if (!error) {
        const modifications = response.value
        return [
          setAll(modifications),
          getFeedsRoutesAndStops({
            bundleId,
            modifications
          })
        ]
      }
    }
  })
}

export const saveToServer = (m: Modification) =>
  fetch({
    url: `/api/modification/${m._id}`,
    options: {
      body: formatForServer(m),
      method: 'put'
    },
    next (error, response) {
      if (!error) {
        return setLocally(response.value)
      }
    }
  })

export const set = (m: Modification) => [setLocally(m), saveToServer(m)]
export const setActive = createAction('set active modification')
const setAll = createAction('set modifications')
export const setAndRetrieveData = ({bundleId, modification}: {
  bundleId: string,
  modification: Modification
}) =>
  (isValid(modification)
    ? [
      set(modification),
      getFeedsRoutesAndStops({bundleId, modifications: [modification]})
    ]
    : [lockUiWithError('Modification is not valid after operation')])

export const setLocally = createAction('set modification')

/**
 * When deleting a modification's timetable it might have been referenced from
 * other timetables. Ensure that those references are cleared.
 */
export function clearReferencesToTimetable (timetableId: string, modifications: Modification[]) {
  const timetableMatches = tt =>
    tt.phaseFromTimetable && tt.phaseFromTimetable.split(':')[1] === timetableId
  const modificationReferencesTimetable = tts =>
    tts && tts.length > 0 && !!tts.find(timetableMatches)
  const clearTimetableReferences = tt =>
    (timetableMatches(tt)
      ? {...tt, phaseFromTimetable: null, phaseFromStop: null}
      : tt)

  return modifications
    .filter(m => modificationReferencesTimetable(m.entries || m.timetables))
    .map(
      m =>
        (m.entries
          ? set({...m, entries: m.entries.map(clearTimetableReferences)})
          : set({...m, timetables: (m.timetables || []).map(clearTimetableReferences)}))
    )
}

/**
 * When deleting a modification it might have been referenced from other
 * modifications. Ensure that those references are cleared.
 */
export const clearReferencesToModification = createAction(
  'clear references to modification',
  ({_id, modifications}) => {
    const timetableMatches = tt =>
      tt.phaseFromTimetable && tt.phaseFromTimetable.split(':')[0] === _id
    const referencesModification = tts =>
      tts && tts.length > 0 && !!tts.find(timetableMatches)
    const clearTimetableReferences = tt =>
      (timetableMatches(tt)
        ? {...tt, phaseFromTimetable: null, phaseFromStop: null}
        : tt)

    return modifications
      .filter(m => referencesModification(m.entries || m.timetables))
      .map(
        m =>
          (m.entries
            ? set({...m, entries: m.entries.map(clearTimetableReferences)})
            : set({
              ...m,
              timetables: m.timetables.map(clearTimetableReferences)
            }))
      )
  }
)
