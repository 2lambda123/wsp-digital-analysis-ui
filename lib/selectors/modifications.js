
export default function selectModifications (state) {
  return state.scenario.modifications
}
