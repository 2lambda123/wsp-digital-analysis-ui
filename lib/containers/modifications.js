import {connect} from 'react-redux'
import {push} from 'react-router-redux'

import {create as createModification, setAndRetrieveData} from '../actions/modifications'
import Modifications from '../components/modification/list'
import activeModificationSelector from '../selectors/active-modification'

function mapStateToProps (state, props) {
  const defaultFeedId = state.scenario.feeds.length > 0
    ? state.scenario.feeds[0].id
    : ''

  return {
    activeModification: activeModificationSelector(state, props),
    bundleId: state.scenario.currentBundle.id,
    defaultFeedId,
    modifications: state.scenario.modifications || [],
    modificationsByType: state.scenario.modificationsByType || {},
    projectId: props.params.projectId,
    scenarioId: props.params.scenarioId,
    variants: state.scenario.variants
  }
}

function mapDispatchToProps (dispatch) {
  return {
    createModification: (opts) => dispatch(createModification(opts)),
    push: (url) => dispatch(push(url)),
    updateModification: (opts) => dispatch(setAndRetrieveData(opts))
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Modifications)
