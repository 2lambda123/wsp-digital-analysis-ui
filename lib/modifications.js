import React, {PropTypes} from 'react'
import {connect} from 'react-redux'

import {create as createModification, setAndRetrieveData} from './actions/modifications'
import DeepEqualComponent from './components/deep-equal'
import activeModificationSelector from './selectors/active-modification'
import * as types from './utils/modification-types'
import ModificationGroup from './modification-group'
import VariantEditor from './variant-editor'

function mapStateToProps (state, props) {
  const defaultFeedId = state.scenario.feeds.length > 0
    ? state.scenario.feeds[0].id
    : ''

  return {
    activeModification: activeModificationSelector(state, props),
    bundleId: state.scenario.currentBundle.id,
    defaultFeedId,
    modificationsByType: state.scenario.modificationsByType || {},
    projectId: props.params.projectId,
    scenarioId: props.params.scenarioId,
    variants: state.scenario.variants
  }
}

function mapDispatchToProps (dispatch) {
  return {
    createModification: (opts) => dispatch(createModification(opts)),
    replaceModification: (opts) => dispatch(setAndRetrieveData(opts))
  }
}

class Modifications extends DeepEqualComponent {
  static propTypes = {
    activeModification: PropTypes.object,
    bundleId: PropTypes.string.isRequired,
    createModification: PropTypes.func.isRequired,
    defaultFeedId: PropTypes.string.isRequired,
    modificationsByType: PropTypes.object.isRequired,
    projectId: PropTypes.string.isRequired,
    replaceModification: PropTypes.func.isRequired,
    scenarioId: PropTypes.string.isRequired,
    variants: PropTypes.array.isRequired
  }

  _createFns = {}

  _createModificationBy (type) {
    if (!this._createFns[type]) {
      this._createFns[type] = () => this._createModification(type)
    }
    return this._createFns[type]
  }

  _createModification (type) {
    const {createModification, defaultFeedId, projectId, scenarioId, variants} = this.props
    createModification({
      feedId: defaultFeedId,
      projectId,
      scenarioId,
      type,
      variants: variants.map((v) => true)
    })
  }

  _replaceModification = (modification) => {
    const {bundleId, replaceModification} = this.props
    replaceModification({
      bundleId,
      modification
    })
  }

  render () {
    const {activeModification, modificationsByType, projectId, scenarioId} = this.props

    return (
      <div>
        <VariantEditor />

        {modificationsByType && Object.values(types).map((type) => {
          return <ModificationGroup
            activeModification={activeModification}
            create={this._createModificationBy(type)}
            key={`modification-group-${type}`}
            modifications={modificationsByType[type] || []}
            projectId={projectId}
            replaceModification={this._replaceModification}
            scenarioId={scenarioId}
            type={type}
            />
        })}
      </div>
    )
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Modifications)
