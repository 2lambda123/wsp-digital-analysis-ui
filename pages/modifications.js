import {loadProjectAndModifications} from 'lib/actions/project'
import Modifications from 'lib/components/modification/list'

/**
 * Populates props with bundle, feeds, modifications, and project.
 */
Modifications.getInitialProps = async ctx => {
  const {projectId} = ctx.query
  return await ctx.reduxStore.dispatch(loadProjectAndModifications(projectId))
}

export default Modifications
