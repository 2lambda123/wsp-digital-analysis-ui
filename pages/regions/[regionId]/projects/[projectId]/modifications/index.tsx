import {getForProject as loadModifications} from 'lib/actions/modifications'
import {loadProject} from 'lib/actions/project'
import List from 'lib/components/modification/list'
import ProjectTitle from 'lib/components/project-title'
import MapLayout from 'lib/layouts/map'
import withInitialFetch from 'lib/with-initial-fetch'

/**
 * Show Select Project if a project has not been selected
 */
const ModificationsPage: any = withInitialFetch(
  ({modifications, project}) => (
    <>
      <ProjectTitle project={project} />
      <List modifications={modifications} project={project} />
    </>
  ),
  async (dispatch, query) => {
    const [project, modifications] = await Promise.all([
      dispatch(loadProject(query.projectId)),
      dispatch(loadModifications(query.projectId))
    ])
    return {project, modifications}
  }
)

ModificationsPage.Layout = MapLayout

export default ModificationsPage
