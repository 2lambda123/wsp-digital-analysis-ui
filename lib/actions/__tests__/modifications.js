/* global describe, expect, it */

import nock from 'nock'

import * as modifications from '../modifications'

const mockModification = {
  feed: { id: '1' },
  id: '1',
  routes: [{ id: '1' }],
  stops: [{ id: '1' }],
  trips: [{ id: '1' }],
  variants: ['2']
}

describe('actions > modifications', () => {
  it('copyFromScenario should work', async () => {
    nock('http://mockHost.com/')
      .get('/api/scenario/1/modifications')
      .reply(200, [mockModification])
      .put(/^\/api\/modification/)
      .reply(200, 'saved')

    const copyResult = modifications.copyFromScenario({
      fromScenarioId: 1,
      toScenarioId: 2,
      variants: [1]
    })

    // expect to receive increment action while making request
    expect(copyResult[0]).toMatchSnapshot()

    // perform request
    const fetchResult = await copyResult[1]

    // expect to receive decrement action upon fulfillment of request
    expect(fetchResult[0]).toMatchSnapshot()

    // parse response
    const parseResult = await fetchResult[1]

    // should receive an array of arrays
    expect(parseResult.length).toBe(1)
    expect(parseResult[0].length).toBe(2)

    // assert existance of uuid then delete it for snapshot
    expect(parseResult[0][0].payload.id).toMatch(/[-\w]*/)
    delete parseResult[0][0].payload.id

    // expect parse result to be the result of the set action
    expect(parseResult[0][0]).toMatchSnapshot()

    // expect a set action
    expect(parseResult[0][1]).toMatchSnapshot()

    // don't bother to perform save request
    // since it is tested in another test case in this file
  })

  it('create should work', async () => {
    nock('http://mockHost.com/')
      .post('/api/modification')
      .reply(200, mockModification)

    const createResult = modifications.create({
      feedId: '1',
      projectId: '1',
      scenarioId: '1',
      type: 'add-trip-pattern',
      variants: []
    })

    // expect to receive increment action while making request
    expect(createResult[0]).toMatchSnapshot()

    // perform request
    const fetchResult = await createResult[1]

    // expect to receive decrement action upon fulfillment of request
    expect(fetchResult[0]).toMatchSnapshot()

    // parse response
    const parseResult = await fetchResult[1]

    // expect parseResult to have server response and
    // react route push action
    expect(parseResult).toMatchSnapshot()
  })

  it('deleteModification should work', async () => {
    nock('http://mockHost.com/')
      .delete('/api/modification/1')
      .reply(200, 'deleted')

    const deleteResult = modifications.deleteModification('1')

    // expect to receive delete locally action
    expect(deleteResult[0]).toMatchSnapshot()

    // perform request
    const fetchResult = await deleteResult[1]

    // expect to receive increment action while making request
    expect(fetchResult[0]).toMatchSnapshot()

    // parse response
    const parseResult = await fetchResult[1]

    // expect to receive decrement action upon fulfillment of request
    expect(parseResult).toMatchSnapshot()
  })

  it('getForScenario should work', async () => {
    nock('http://mockHost.com/')
      .get('/api/scenario/1/modifications')
      .reply(200, [mockModification])
      .get(/^\/api\/graphql/)
      .reply(200)

    const getResult = modifications.getForScenario({
      bundleId: '1',
      scenarioId: '1'
    })

    // expect to receive increment action while making request
    expect(getResult[0]).toMatchSnapshot()

    // perform request
    const fetchResult = await getResult[1]

    // expect to receive decrement action upon fulfillment of request
    expect(fetchResult[0]).toMatchSnapshot()

    // parse response
    const parseResult = await fetchResult[1]

    // expect to receive an array with first result being a setAll action
    expect(parseResult[0]).toMatchSnapshot()

    // expect second result being a request to getFeedsRoutesAndStops
    expect(parseResult[1]).toMatchSnapshot()

    // don't bother testing result of getFeedsRoutesAndStops
    // since it is tested in the get-feeds-routes-and-stops file
  })

  it('saveToServer should work', async () => {
    nock('http://mockHost.com/')
      .put('/api/modification/1')
      .reply(200, 'saved')

    const saveResult = modifications.saveToServer(mockModification)

    // expect to receive increment action while making request
    expect(saveResult[0]).toMatchSnapshot()

    // perform request
    const fetchResult = await saveResult[1]

    // expect to receive only decrement action upon fulfillment of request
    expect(fetchResult.length).toBe(1)
    expect(fetchResult[0]).toMatchSnapshot()
  })

  it('set should work', async () => {
    nock('http://mockHost.com/')
      .put('/api/modification/1')
      .reply(200, 'saved')

    const setResult = modifications.set(mockModification)

    // should receive an array of arrays
    expect(setResult.length).toBe(2)

    // expect a setLocally action
    expect(setResult[0]).toMatchSnapshot()

    // expect saveToServer action
    expect(setResult[1]).toMatchSnapshot()

    // don't bother testing result of saveToServer
    // since it is tested in another test case in this file
  })

  it('setActive should work', () => {
    expect(modifications.setActive()).toMatchSnapshot()
  })

  it('setAndRetrieveData should work', async () => {
    nock('http://mockHost.com/')
      .put('/api/modification/1')
      .reply(200, 'saved')
      .get(/^\/api\/graphql/)
      .reply(200)

    const setResult = modifications.setAndRetrieveData({
      bundleId: 1,
      modification: mockModification
    })

    // should receive an array
    expect(setResult.length).toBe(2)

    // expect a set action (array of setLocally and saveToServer)
    expect(setResult[0]).toMatchSnapshot()

    // expect a getFeedsRoutesAndStops action
    expect(setResult[1]).toMatchSnapshot()

    // don't bother testing result of getFeedsRoutesAndStops or saveToServer
    // getFeedsRoutesAndStops is tested in the get-feeds-routes-and-stops file
    // saveToServer is tested in another test case in this file
  })

  it('setLocally should work', () => {
    expect(modifications.setLocally()).toMatchSnapshot()
  })
})
