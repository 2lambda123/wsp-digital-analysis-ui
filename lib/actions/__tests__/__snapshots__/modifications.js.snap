exports[`actions > modifications copyFromScenario should work 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > modifications copyFromScenario should work 2`] = `
Object {
  "type": "decrement outstanding requests",
}
`;

exports[`actions > modifications copyFromScenario should work 3`] = `
Object {
  "payload": Object {
    "bidirectional": false,
    "entries": Array [],
    "feed": "1",
    "name": "Test Modification",
    "routes": Array [
      "route1",
    ],
    "scenario": 2,
    "segments": Array [],
    "showOnMap": false,
    "timetables": Array [
      Object {
        "dwellTime": 10,
        "endTime": 57600,
        "friday": true,
        "headwaySecs": 900,
        "monday": true,
        "name": "Test timetable",
        "patternTrips": Array [
          "abcd",
        ],
        "saturday": false,
        "segmentSpeeds": Array [],
        "sourceTrip": "abcd",
        "startTime": 28800,
        "sunday": false,
        "thursday": true,
        "tuesday": true,
        "wednesday": true,
      },
    ],
    "trips": Array [
      "abcd",
    ],
    "type": "add-trip-pattern",
    "variants": Array [
      0,
    ],
  },
  "type": "set modification",
}
`;

exports[`actions > modifications copyFromScenario should work 4`] = `
Array [
  Object {
    "type": "increment outstanding requests",
  },
  Promise {},
]
`;

exports[`actions > modifications create should work 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > modifications create should work 2`] = `
Object {
  "type": "decrement outstanding requests",
}
`;

exports[`actions > modifications create should work 3`] = `
Array [
  Object {
    "payload": Object {
      "bidirectional": false,
      "entries": Array [],
      "feed": "1",
      "id": "1234",
      "name": "Test Modification",
      "routes": Array [
        "route1",
      ],
      "segments": Array [],
      "showOnMap": false,
      "timetables": Array [
        Object {
          "dwellTime": 10,
          "endTime": 57600,
          "friday": true,
          "headwaySecs": 900,
          "monday": true,
          "name": "Test timetable",
          "patternTrips": Array [
            "abcd",
          ],
          "saturday": false,
          "segmentSpeeds": Array [],
          "sourceTrip": "abcd",
          "startTime": 28800,
          "sunday": false,
          "thursday": true,
          "tuesday": true,
          "wednesday": true,
        },
      ],
      "trips": Array [
        "abcd",
      ],
      "type": "add-trip-pattern",
      "variants": Array [
        0,
      ],
    },
    "type": "set modification",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/projects/1/scenarios/1/modifications/1234",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;

exports[`actions > modifications deleteModification should work 1`] = `
Object {
  "payload": "1",
  "type": "delete modification",
}
`;

exports[`actions > modifications deleteModification should work 2`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > modifications deleteModification should work 3`] = `
Array [
  Object {
    "type": "decrement outstanding requests",
  },
]
`;

exports[`actions > modifications getForScenario should work 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > modifications getForScenario should work 2`] = `
Object {
  "type": "decrement outstanding requests",
}
`;

exports[`actions > modifications getForScenario should work 3`] = `
Object {
  "payload": Array [
    Object {
      "bidirectional": false,
      "entries": Array [],
      "feed": "1",
      "id": "1234",
      "name": "Test Modification",
      "routes": Array [
        "route1",
      ],
      "segments": Array [],
      "showOnMap": false,
      "timetables": Array [
        Object {
          "dwellTime": 10,
          "endTime": 57600,
          "friday": true,
          "headwaySecs": 900,
          "monday": true,
          "name": "Test timetable",
          "patternTrips": Array [
            "abcd",
          ],
          "saturday": false,
          "segmentSpeeds": Array [],
          "sourceTrip": "abcd",
          "startTime": 28800,
          "sunday": false,
          "thursday": true,
          "tuesday": true,
          "wednesday": true,
        },
      ],
      "trips": Array [
        "abcd",
      ],
      "type": "add-trip-pattern",
      "variants": Array [
        0,
      ],
    },
  ],
  "type": "set modifications",
}
`;

exports[`actions > modifications getForScenario should work 4`] = `
Array [
  Object {
    "type": "increment outstanding requests",
  },
  Promise {},
]
`;

exports[`actions > modifications saveToServer should work 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > modifications saveToServer should work 2`] = `
Object {
  "type": "decrement outstanding requests",
}
`;

exports[`actions > modifications set should work 1`] = `
Object {
  "payload": Object {
    "bidirectional": false,
    "entries": Array [],
    "feed": "1",
    "id": "1234",
    "name": "Test Modification",
    "routes": Array [
      "route1",
    ],
    "segments": Array [],
    "showOnMap": false,
    "timetables": Array [
      Object {
        "dwellTime": 10,
        "endTime": 57600,
        "friday": true,
        "headwaySecs": 900,
        "monday": true,
        "name": "Test timetable",
        "patternTrips": Array [
          "abcd",
        ],
        "saturday": false,
        "segmentSpeeds": Array [],
        "sourceTrip": "abcd",
        "startTime": 28800,
        "sunday": false,
        "thursday": true,
        "tuesday": true,
        "wednesday": true,
      },
    ],
    "trips": Array [
      "abcd",
    ],
    "type": "add-trip-pattern",
    "variants": Array [
      0,
    ],
  },
  "type": "set modification",
}
`;

exports[`actions > modifications set should work 2`] = `
Array [
  Object {
    "type": "increment outstanding requests",
  },
  Promise {},
]
`;

exports[`actions > modifications setActive should work 1`] = `
Object {
  "type": "set active modification",
}
`;

exports[`actions > modifications setAndRetrieveData should work 1`] = `
Array [
  Object {
    "payload": Object {
      "bidirectional": false,
      "entries": Array [],
      "feed": "1",
      "id": "1234",
      "name": "Test Modification",
      "routes": Array [
        "route1",
      ],
      "segments": Array [],
      "showOnMap": false,
      "timetables": Array [
        Object {
          "dwellTime": 10,
          "endTime": 57600,
          "friday": true,
          "headwaySecs": 900,
          "monday": true,
          "name": "Test timetable",
          "patternTrips": Array [
            "abcd",
          ],
          "saturday": false,
          "segmentSpeeds": Array [],
          "sourceTrip": "abcd",
          "startTime": 28800,
          "sunday": false,
          "thursday": true,
          "tuesday": true,
          "wednesday": true,
        },
      ],
      "trips": Array [
        "abcd",
      ],
      "type": "add-trip-pattern",
      "variants": Array [
        0,
      ],
    },
    "type": "set modification",
  },
  Array [
    Object {
      "type": "increment outstanding requests",
    },
    Promise {},
  ],
]
`;

exports[`actions > modifications setAndRetrieveData should work 2`] = `
Array [
  Object {
    "type": "increment outstanding requests",
  },
  Promise {},
]
`;

exports[`actions > modifications setLocally should work 1`] = `
Object {
  "type": "set modification",
}
`;
